### ISSUE TITLE
Keg Screen 'tapinfo Report' Fails to POST

### ISSUE THREAD

**Issue by austinpe** [lbussy/keg-cop#20]
I'm using the Keg Screen functionality as a sort of 'web hook' to allow for reactive updates to an LCD screen instead of constantly polling the API.

I'm mostly using the temp and pour reports for my LCD, but while testing I noticed that the tapinfo report was always failing.

Here's a serial log of what I see. You can see I navigate to the taps page and toggle a tap to be active.
This triggers the tap report POST, but it fails.

> 2022-06-21T14:21:06Z V: Processing post to /api/v1/config/taps/.
> 2022-06-21T14:21:06Z V: Checking handleTapPost.
> 2022-06-21T14:21:06Z V: Processing [tap]:(1) pair.
> 2022-06-21T14:21:06Z N: Settings update, processing [tap]:(1).
> 2022-06-21T14:21:06Z V: Processing [ppu]:(21120) pair.
> 2022-06-21T14:21:06Z N: Settings update, [ppu]:(21120) applied.
> 2022-06-21T14:21:06Z V: Processing [beername]:(Test1) pair.
> 2022-06-21T14:21:06Z N: Settings update, [beername]:(Test1) applied.
> 2022-06-21T14:21:06Z V: Processing [cap]:(5.0000) pair.
> 2022-06-21T14:21:06Z N: Settings update, [cap]:(5.0000) applied.
> 2022-06-21T14:21:06Z V: Processing [remain]:(4.8000) pair.
> 2022-06-21T14:21:06Z N: Settings update, [remain]:(4.8000) applied.
> 2022-06-21T14:21:06Z V: Processing [active]:(active) pair.
> 2022-06-21T14:21:06Z N: Settings update [1], [active]:(active) applied.
> 2022-06-21T14:21:06Z W: Warning: Previous transaction for Taps POST is in an unknown state (0).
> 2022-06-21T14:21:06Z V: Checking handleTapCal.
> 2022-06-21T14:21:06Z V: Processing [tap]:(1) pair.
> 2022-06-21T14:21:06Z V: Processing [ppu]:(21120) pair.
> 2022-06-21T14:21:06Z V: Processing [beername]:(Test1) pair.
> 2022-06-21T14:21:06Z V: Processing [cap]:(5.0000) pair.
> 2022-06-21T14:21:06Z V: Processing [remain]:(4.8000) pair.
> 2022-06-21T14:21:06Z V: Processing [active]:(active) pair.
> 2022-06-21T14:21:07Z V: Checking handleSetCalMode.
> 2022-06-21T14:21:07Z V: Clearing any calibration flags before setting new flags.
> 2022-06-21T14:21:07Z V: Processing [tap]:(1) pair.
> 2022-06-21T14:21:07Z V: Processing [ppu]:(21120) pair.
> 2022-06-21T14:21:07Z V: Processing [beername]:(Test1) pair.
> 2022-06-21T14:21:07Z V: Processing [cap]:(5.0000) pair.
> 2022-06-21T14:21:07Z V: Processing [remain]:(4.8000) pair.
> 2022-06-21T14:21:07Z V: Processing [active]:(active) pair.
> 2022-06-21T14:21:07Z V: MQTT: No broker configured.

From my testing the pour, temp, and cooling reports all POST just fine. 
I have not tested the kick report.

**Comment by lbussy** [comment]
Does it work properly for you when you toggle active/inactive via the normal web interface?

**Comment by austinpe** [comment]
This was via the normal web interface.
I have not tried to toggle it via API but I'll try that now.

**Comment by lbussy** [comment]
I believe this is also fixed in Alpha.2, give me a few to upload and (pre)release.

**Comment by austinpe** [comment]
Sadly I still get the following in the Serial Out when I update a tap.

> 2022-06-24T21:34:13Z V: Processing post to /api/v1/config/taps/.
> 2022-06-24T21:34:13Z V: Checking handleTapPost.
> 2022-06-24T21:34:13Z V: Processing [tap]:(0) pair.
> 2022-06-24T21:34:13Z N: Settings update, processing [tap]:(0).
> 2022-06-24T21:34:13Z V: Processing [label]:(1) pair.
> 2022-06-24T21:34:13Z N: Settings update, processing [label]:(1).
> 2022-06-24T21:34:13Z V: Processing [ppu]:(21120) pair.
> 2022-06-24T21:34:13Z N: Settings update, [ppu]:(21120) applied.
> 2022-06-24T21:34:13Z V: Processing [bevname]:(Pudswiller Doors) pair.
> 2022-06-24T21:34:13Z N: Settings update, [bevname]:(Pudswiller Doors) applied.
> 2022-06-24T21:34:13Z V: Processing [cap]:(5.0000) pair.
> 2022-06-24T21:34:13Z N: Settings update, [cap]:(5.0000) applied.
> 2022-06-24T21:34:13Z V: Processing [remain]:(4.2000) pair.
> 2022-06-24T21:34:13Z N: Settings update, [remain]:(4.2000) applied.
> 2022-06-24T21:34:13Z V: Processing [active]:(active) pair.
> 2022-06-24T21:34:13Z N: Settings update [0], [active]:(active) applied.
> 2022-06-24T21:342022-06-24T21:34:13Z V: Checking handleTapCal.
> 2022-06-24T21:34:13Z V: Processing [tap]:(0) pair.
> 2022-06-24T21:34:13Z V: Processing [label]:(1) pair.
> 2022-06-24T21:34:13Z V: Processing [ppu]:(21120) pair.
> 2022-06-24T21:34:13Z V: Processing [bevname]:(Pudswiller Doors) pair.
> 2022-06-24T21:34:13Z V: Processing [cap]:(5.0000) pair.
> 2022-06-24T21:34:13Z V: Processing [remain]:(4.2000) pair.
> 2022-06-24T21:34:13Z V: Processing [active]:(active) pair.
> 2022-06-24T21:34:13Z V: Checking handleSetCalMode.
> 2022-06-24T21:34:13Z V: Clearing any calibration flags before setting new flags.
> 2022-06-24T21:34:14Z V: Processing [tap]:(0) pair.
> 2022-06-24T21:34:14Z V: Processing [label]:(1) pair.
> 2022-06-24T21:34:14Z V: Processing [ppu]:(21120) pair.
> 2022-06-24T21:34:14Z V: Processing [bevname]:(Pudswiller Doors) pair.
> 2022-06-24T21:34:14Z V: Processing [cap]:(5.0000) pair.
> 2022-06-24T21:34:14Z V: Processing [remain]:(4.2000) pair.
> 2022-06-24T21:34:14Z V: Processing [active]:(active) pair.
> :13Z W: Warning: Previous transaction for Taps POST is in an unknown state (0).

I confirm I still am getting all other reports, just not the tapinfo report.
Here's a successful temp report:

> 2022-06-24T21:36:53Z V: Temps: Preparing POST to: http://192.168.0.27:8080/
> 2022-06-24T21:36:53Z V: Temps POST to http://192.168.0.27:8080/ dispatched.
> 2022-06-24T21:36:53Z N: Temps: HTTP response code 200 received (81ms): Ok.

**Comment by lbussy** [comment]
Can you try disabling all browser extensions and/or using private mode?   I can't reproduce that at all.

**Comment by austinpe** [comment]
I tested in both an incognito tab with all extensions disabled AND with Postman as a pure API PUT request.

Both times did not trigger a POST received by the listening server. Temp reports are actively being received. 

I was not able to listen to the serial out yet. 
I tried to flash a 2nd esp32 as a dummy, but it did not let me monitor serial out and use the web page at the same time?
If I had the webpage loaded and opened PuTTY it would show that it caused a panic? Weird. I'll investigate that more later.

**Comment by lbussy** [comment]
OH!  I'm sorry I was being dense.  Now I get it.  I'll need to test that again.

**Comment by lbussy** [comment]
Just dropping notes here.  I tested with brewpi (`brewpi-api.php`) and I was able to receive the JSON:

```
2022-06-25 07:06:24  Received JSON: {"api":"Keg Cop","guid":"0EF7C6300000XXXX","hostname":"kegcop","breweryname":"Silver Fox Brewery","kegeratorname":"Keezer","reporttype":"tempreport","imperial":true,"controlpoint":4,"setting":35,"status":6,"controlenabled":true,"sensors":[{"name":"Room","value":-196.6000061,"enabled":true},{"name":"Tower","value":-196.6000061,"enabled":true},{"name":"Upper Chamber","value":-196.6000061,"enabled":true},{"name":"Lower Chamber","value":-196.6000061,"enabled":true},{"name":"Keg","value":-196.6000061,"enabled":true}]}
```

I did also receive the error on the terminal:

```
2022-06-25T12:22:17Z W: Warning: Previous transaction for Taps POST is in an unknown state (0).
```

So, I'm with you so far.  All of this would be unnecessary if @thorrak would get Keg Screen done.  😉 

Anyway now I know what you are seeing.  I'll poke at it.

**Comment by austinpe** [comment]
@thorrak , as side note / plug, I see you've been using Vue on your webapps. I'm a Vue dev and happy to help or test if ya need it.

**Comment by lbussy** [comment]
>> I see you've been using Vue on your webapps.

"Using."  It's using him.  😃

**Comment by thorrak** [comment]
> > I see you've been using Vue on your webapps.
> 
> "Using." It's using him. 😃

^^ THIS. 🥲

I'll definitely take you up on that offer @austinpe 

I've been mostly enjoying Vue as I've been learning it (BrewFlasher Web, TronBridge) but have run into other projects that it's been giving me fits with (KegScreen's tap lists, Fermentrack's dashboard). If I had more time (and fewer projects) I've contemplated rewriting Fermentrack to be mostly REST + Vue rather than relying on Django templates, though I'll probably end up doing the same for KegScreen first.

**Comment by austinpe** [comment]
Yeah, from my experience Django feels really 'heavy' vs just having the routing handled by Vue.
There are great, well tested libraries to handle REST within JS such as [Axios](https://github.com/axios/axios).

Obviously, we all use the tools we're most comfortable with and getting a project out and useable is the main goal.

My go to recently has been Vue3 + Tailwind; it has allowed me to build awesome projects very quickly and without being a designer.
I made two local brewery's websites with that combo: [King Maker Brewing](https://www.kingmakerbrewing.com/) and [Reve](https://www.revebrewing.com/)

But yeah, redoing Fermentrack to use Vue + Rest vs Django would possibly make it easier to maintain, but would be a ton of work.

**Comment by lbussy** [comment]
Where I generally try to c-block John getting any attention, in this case him finishing KegScreen benefits me directly so I approve. 😛

**Comment by lbussy** [comment]
Here ya go:  https://web.brewflasher.com/fw/78/

(in case it's broken again, it's under pre-releases Alpha.3.)

**Comment by lbussy** [comment]
Any issues with this one @austinpe?

**Comment by lbussy** [comment]
Well, I think it works now. :)  This was closed by #25 and merged into `devel`.

I'd like to get the TaplistIO work done and then release 1.1.0.

**Comment by austinpe** [comment]
Good news and bad news again.

**Good news**
I confirm that the tap info report successfully sends now! 🥳 

**Bad news**
I'm not sure what happened, but as of these last few builds I can no longer monitor serial out.
As soon as I connect via PuTTY it causes a panic (see log below).
Also, rebooting doesn't fix it. I have to hold the 'boot' button for a few seconds before it will properly reload the firmware.

If I don't connect on the serial line and just let it go then it works fine and I can interact with it via the webpage / API.

> rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
> configsip: 0, SPIWP:0xee
> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
> mode:DIO, clock div:2
> load:0x3fff0018,len:4
> load:0x3fff001c,len:1044
> load:0x40078000,len:10044
> load:0x40080400,len:5872
> entry 0x400806ac
> 
> 1970-01-01T00:00:00Z N: Serial logging started at 115200.
> 1970-01-01T00:00:00Z V: Starting WiFi.
> *WM: [1] Free heap:        246792
> *WM: [1] ESP SDK version:  v3.3.5-1-g85c43024c
> *WM: [3] allocating params bytes: 20
> *WM: [2] Added Parameter: name
> *WM: [1] AutoConnect.
> *WM: [2] esp_wifi_set_country: US
> *WM: [2] ESP32 event handler enabled.
> *WM: [2] Connecting as WiFi client.
> *WM: [3] STA static IP:
> *WM: [2] setSTAConfig static ip not set, skipping.
> *WM: [1] Connecting to saved AP: REDACTED
> *WM: [3] Using password: REDACTED
> *WM: [3] WiFi_enableSTA:  enabled.
> *WM: [3] Mode after delay: STA
> *WM: [2] 30000 ms connectTimeout set.
> *WM: ▒▒] 30000 ms timeout, waiting for connection................Guru Meditation Error: Core  panic'ed (Interrupt wdt timeout on CPU0)
> Core 0 register dump:
> PC       0x40086e6  PS       0x00060834  A0       0x8012c10  A1       0x3fa3e
> A2       0x3fc604  A3       0x00000000  A4       0x00000000  A5       0x3fcd0
> A6       0x00000002  A7       0x00000100  A8       0x8008a44  A9       0x3fa3f
> A10      0x00000005  A11      0x00060c3  A12      0x00060c0  A13      0x3fa488
> A14      0x00000017  A15      0x00000000  SAR      0x00000020  EXCCAUSE 0x00000005
> EXCVADDR 0x00000000  LBEG     0x4000ce  LEND     0x4000cf  LCOUNT   0xffff
> 
> ELF file SHA256: 000000000000000
> 
> Backtrace:0x40086e6:0x3fa3e 0x4012c0d0x3fa420 0x4012c19:0x3fa450 0x40130ff0x3fa480 0x40089be0x3fa4b
> 
> ore 1 register dump:
> PC       0x4018762a PS       0x00060834  A0       0x8016d1e A1       0x3fb7a
> A2       0x00000000  A3       0x00000001  A4       0x80089d0  A5       0x3fba0
> A6       0x00000000  A7       0x3fb4a  A8       0x00060023  A9       0x3fb310
> A10      0x00000000  A11      0x00000001  A12      0x80089d0  A13      0x3fb8b
> A14      0x00000000  A15      0x3fb4a  SAR      0x00000000  EXCCAUSE 0x00000005 
> EXCVADDR 0x00000000  LBEG     0x00000000  LEND     0x00000000  LCOUNT   0x00000000
> 
> ELF file SHA256: 000000000000000
> 
> Backtrace:0x4018762a0x3fb7a 0x4016d1b0x3fb7c 0x4008b7d0x3fb7e 0x40089be0x3fb800
> 
> Rebooting...

**Comment by lbussy** [comment]
> I'm not sure what happened, but as of these last few builds I can no longer monitor serial out.

Opened #26 for this.

I monitor it via Serial, so it's very curious why you cannot.
